### PR TITLE
Extend query depth limiter allowing for more detailed rules 

### PR DESCRIPTION
**Pull request by tsmith023**
This PR enhances the query depth limiter by providing options to specify additional rules for fields within the GraphQL query.
These can be any or all of the field name, the field arguments, and the field keys.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

- https://github.com/strawberry-graphql/strawberry/issues/2338

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tsmith023** [comment]
This seems to have included some commits from a previous PR. Hopefully this won't cause a problem! :grin:

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This change allows clients to define connectionParams when making Subscription requests similar to the way [Apollo-Server](https://www.apollographql.com/docs/apollo-server/data/subscriptions/#operation-context) does it.

With [Apollo-Client (React)](https://www.apollographql.com/docs/react/data/subscriptions/#5-authenticate-over-websocket-optional) as an example, define a Websocket Link:
```
import { GraphQLWsLink } from '@apollo/client/link/subscriptions';
import { createClient } from 'graphql-ws';

const wsLink = new GraphQLWsLink(createClient({
  url: 'ws://localhost:4000/subscriptions',
  connectionParams: {
    authToken: user.authToken,
  },
}));
```
and the JSON passed to `connectionParams` here will appear within Strawberry's context as the `connection_params` attribute when accessing `info.context` within a Subscription resolver.


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Tommy Smith for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/extensions/query_depth_limiter.py [hunk 1]
```diff
@@ -220,18 +236,62 @@ def is_ignored(node: FieldNode, ignore: Optional[List[IgnoreType]] = None) -> bo
...
+            return True
+    elif isinstance(rule, re.Pattern):
+        if rule.match(field_name):
+            return True
+    elif callable(rule):
+        if rule(field_name):
+            return True
+    else:
+        raise ValueError(f"Invalid ignore option: {rule}")
+
```
Comment (patrick91) [inline strawberry/extensions/query_depth_limiter.py:271]: do you think we can add coverage for these lines? :)

Comment (tsmith023) [reply strawberry/extensions/query_depth_limiter.py:271]: With the codecov report, I'll go through the additions to ensure it's all covered!

**Comment by tsmith023** [comment]
I'm going to re-create my fork to merge from since the commit history seems to be broken here. Will reopen this PR once it's all sorted!

**Pull request closed without merging**